### PR TITLE
update C queries and grammar

### PR DESCRIPTION
**Pull request by kirawi**
The updated queries were pulled from nvim-treesitter. I might make another PR to make it more accurate in the future.

### CODE REVIEW

File: runtime/queries/c/highlights.scm [hunk 1]
```diff
@@ -1,72 +1,148 @@
...
+  "struct"
+  "typedef"
+  "union"
+  "volatile"
+  "goto"
+  "register"
+] @keyword
+
+"sizeof" @keyword.operator
+"return" @keyword.return
```
Comment (archseer) [inline runtime/queries/c/highlights.scm:19]: @keyword.control.return

File: runtime/queries/c/highlights.scm [hunk 2]
```diff
@@ -1,72 +1,148 @@
...
+(comma_expression [ "," ] @operator)
+
+[
+  (true)
+  (false)
+] @constant.builtin.boolean
+
+[
+  (string_literal)
+  (system_lib_string)
```
Comment (archseer) [inline runtime/queries/c/highlights.scm:108]: Second one is `@string.special.path`

File: runtime/queries/c/highlights.scm [hunk 3]
```diff
@@ -76,17 +152,31 @@
...
+  "_unaligned"
+  "__unaligned"
+  "__declspec"
+  (attribute_declaration)
+] @attribute
 
-(comment) @comment
+(field_identifier) @variable.other.member
+
+(ERROR) @error
```
Comment (archseer) [inline runtime/queries/c/highlights.scm:182]: I've seen this added more and more, I've had this added to highlights automatically before: https://github.com/helix-editor/helix/blob/c2e41082e4b2e0e1dba6a980a7f09ccaad88cbd0/helix-core/src/syntax.rs#L359-L360

I think we should remove this everywhere, use the snippet above and make it a config option.

Comment (kirawi) [reply runtime/queries/c/highlights.scm:182]: Will it cause the query to fail if an `ERROR` node isn't defined for the grammar?

Comment (archseer) [reply runtime/queries/c/highlights.scm:182]: The `ERROR` node is a meta node automatically inserted by tree-sitter on a parsing error.

File: book/src/languages.md [hunk 4]
```diff
@@ -61,6 +61,7 @@ These configuration keys are available:
 | `config`              | Language Server configuration                                 |
 | `grammar`             | The tree-sitter grammar to use (defaults to the value of `name`) |
 | `formatter`           | The formatter for the language, it will take precedence over the lsp when defined. The formatter must be able to take the original file as input from stdin and write the formatted file to stdout |
+| `append-error-node`   | Whether to append a `(ERROR) @error` capture to the language query. (defaults to `true`) |
```
Comment (the-mikedavis) [inline book/src/languages.md:64]: For the option name, I think `highlight-errors` or something similar would be more intuitive.

I'm also curious whether this should be

* language-specific vs editor-wide
* on/off by default

IMO the error highlighting is pretty noisy when you're typing out a block but I'm not sure how others feel.

Comment (kirawi) [reply book/src/languages.md:64]: There doesn't seem to be a straightforward way to use the editor config. I could either pass it a parameter (requires changing a lot of calls), or find a way to reference the editor config within the highlighter config.

Comment (the-mikedavis) [reply book/src/languages.md:64]: Hmm yeah good point. Using language config is ok then I think 👍

**Comment by the-mikedavis** [comment]
Now that the C/C++ queries have been updated in https://github.com/helix-editor/helix/pull/4079, this might be closable? The appending error node changes could be split off to a separate PR or this PR could pivot to focusing on that

**Comment by kirawi** [comment]
I should probably make a new PR instead if I do that.

**Pull request closed without merging**